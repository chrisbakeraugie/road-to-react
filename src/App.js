// import logo from './logo.svg';
import './App.css';
import React from 'react';
import { resolve } from 'dns';

//Custom hooks example - naming conventions for hooks; start with 'use'
const useSemiPersistentState = (key, initialState) => {
  const [value, setValue] = React.useState(localStorage.getItem(key) || initialState);

  /**
   * This React Hook is basically saying that whenever, wherever, searchTerm is updated through
   * setSearchTerm, localStorage will always be in sync
   */
  React.useEffect(() => {
    localStorage.setItem(key, value);
  }, [value, key]);
  return [value, setValue];
}

const initialStories = [
  {
    title: 'React',
    url: 'https://reactjs.org/', author: 'Jordan Walke', num_comments: 3,
    points: 4,
    objectID: 0,
  },
  {
    title: 'Redux',
    url: 'https://redux.js.org/', author: 'Dan Abramov, Andrew Clark', num_comments: 2,
    points: 5,
    objectID: 1,
  }
];

/**
 * This simulates fetching data from an API by creating a Promise
 * with an intentionally delayed resolution to practice rendering 
 * data asynchronously using React Hooks
 */
const getAsyncStories = () => (
  // return (Promise.resolve({ data: { stories: initialStories } }));
  new Promise(resolve => (
    setTimeout(
      () => resolve({ data: { stories: initialStories } })
      , 2300)
  ))
);

// Can be used to create more "declarative programming", instead of "imperative programming"
// It is now in control of more predictable state transitions, as they are all managed in one spot
const storiesReducer = (state, action) => {
  switch (action.type) {
    case 'STORIES_FETCH_INIT':
      return {
        ...state,
        isLoading: true,
        isError: false
      };
    case 'STORIES_FETCH_SUCCESS':
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload
      };
    case 'STORIES_FETCH_FAILURE':
      return {
        ...state,
        isLoading: false,
        isError: true
      };
    case 'REMOVE_STORY':
      return {
        ...state,
        data: state.data.filter(story => (
          action.payload.objectID !== story.objectID
        ))
      };
    default: throw new Error();

  }
  // if (action.type === 'SET_STORIES') {
  //   return action.payload;
  // } else if (action.type === 'REMOVE_STORY') {
  //   return (state.filter(story => (
  //     action.payload.objectID !== story.objectID
  //   )))
  // }
  // else {
  //   throw new Error();
  // }
}

const App = () => {

  const [searchTerm, setSearchTerm] = useSemiPersistentState('search', 'React');

  // Replaced "useState" hook with a reducer, which can be used to handle more complicated state management
  // Then, replaced the multiple state hooks with ONE, to help manage/prevent "Impossible States"
  const [stories, dispatchStories] = React.useReducer(storiesReducer,
    { data: [], isLoading: false, isError: false });
  // const [isLoading, setIsLoading] = React.useState(false);
  // const [isError, setIsError] = React.useState(false); // Error handling if third party API errors

  React.useEffect(() => {
    dispatchStories({ type: 'STORIES_FETCH_INIT' });

    getAsyncStories().then(result => {
      dispatchStories({
        type: "STORIES_FETCH_SUCCESS",
        payload: result.data.stories
      });
    }).catch(() => {
      dispatchStories({ type: 'STORIES_FETCH_FAILURE' });
    })
    // setIsLoading(true);
    // getAsyncStories().then(result => {
    //   dispatchStories({
    //     type: 'SET_STORIES',
    //     payload: result.data.stories
    //   });
    //   setIsLoading(false);
    // }).catch(() => {
    //   setIsError(true);
    // });
  }, [])

  // Removed the "removeStory" logic from the handler to the reducer
  const handleRemoveStory = item => {
    dispatchStories({
      type: 'REMOVE_STORY',
      payload: item
    });
  }

  const searchedStories = stories.data.filter((story) => {
    return story.title.toLowerCase().includes(searchTerm.toLowerCase());
  })

  const handleSearch = (event) => {
    setSearchTerm(event.target.value);
    localStorage.setItem('search', event.target.value);
  }

  return (
    <div className="App">

      <h1>Hacker Stories</h1>

      <InputWithLabel
        id="search"
        value={searchTerm}
        isFocused
        onInputChange={handleSearch}
      >
        <strong>Search:</strong>
      </InputWithLabel>
      <hr />

      {stories.isError && <p>Something went wrong...</p>}

      {stories.isLoading ? (
        <p>Loading...</p>
      ) : <List list={searchedStories} onRemoveItem={handleRemoveStory} />
      }
    </div>
  );
}

// We can give the default value for type, but it can still be changed from outside
const InputWithLabel = ({ id, value, type = 'text', isFocused, onInputChange, children }) => {
  // Create a 'ref' with React's "useRef" hook
  // This is a persistent value over the lifecycle of a React component.
  const inputRef = React.useRef();

  // Use a hook into the lifecycle to focus when the component renders
  React.useEffect(() => {
    if (isFocused && inputRef.current) {
      // Access the ref attribute and set the focus
      inputRef.current.focus();
    }
  }, [isFocused])
  return (<div>
    <label htmlFor={id}>{children}</label>
    &nbsp;
    {/* JSX-reserved 'ref' attribute will get the inputRef value */}
    <input ref={inputRef} id={id} type={type} value={value} onChange={onInputChange} autoFocus={isFocused}></input>
  </div>)

};


const List = ({ list, onRemoveItem }) =>
  list.map(item => (
    <Item
      key={item.objectID}
      item={item}
      onRemoveItem={onRemoveItem}
    />
  ));

const Item = ({ item, onRemoveItem }) => {
  const handleRemoveItem = () => {
    onRemoveItem(item);
  }

  return (
    <div>
      <span>
        <a href={item.url}>{item.title}</a>
      </span>
      <span>{item.author}</span>
      <span>{item.num_comments}</span>
      <span>{item.points}</span>
      <span>
        {/* This first method of handler is recommended because: \
          1. It is easier to read (method is above anyway)
          2. It is easier to debug*/}
        <button type="button" onClick={onRemoveItem.bind(null, item)}>Dismiss</button>
        {/* This second method of handler is also acceptable, but is slightly harder to debug */}
        <button type="button" onClick={() => onRemoveItem(item)}>Dismiss (inline)</button>
        {/* This third method is never recommended */}
        <button type="button" onClick={() => {
          onRemoveItem(item);
        }}>Dismiss (wholly defined inline)</button>
      </span>
    </div>
  )
}
export default App;
