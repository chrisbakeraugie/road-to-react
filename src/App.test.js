// create-react-app includes these necessary functions for component testing
import { render, screen, fireEvent, act } from '@testing-library/react';
import App, { storiesReducer, Item, List, SearchForm, InputWithLabel } from './App';
import { expression } from '@babel/template';
import { notDeepStrictEqual } from 'assert';


// the describe() is our TEST SUITE
describe('something truthy and falsy', () => {

  // test() is our TEST CASE
  test('true to be true', () => {
    // expect(true).toBe(true);
    expect(true).toBeTruthy();
  });

  test('false to be false', () => {
    // expect(false).toBe(false);
    expect(false).toBeFalsy();
  })
})

describe('App component', () => {
  test('removes an item when clicking the Dismiss button', () => {

  });

  test('requests some intial stories from an API', () => {

  });
})

// ---------------------- UNIT TESTING: FUNCTIONS ---------------------- //

// Just made up test data
const storyOne = {
  title: 'React',
  url: 'https://reactjs.org/', author: 'Jordan Walke', num_comments: 3,
  points: 4,
  objectID: 0,
};

const storyTwo = {
  title: 'Redux',
  url: 'https://redux.js.org/', author: 'Dan Abramov, Andrew Clark', num_comments: 2,
  points: 5,
  objectID: 1,
};

const stories = [storyOne, storyTwo];

describe('storiesReducer', () => {
  test('removes a story from all stories', () => {
    const action = { type: 'REMOVE_STORY', payload: storyOne };
    const state = { data: stories, isLoading: false, isError: false };

    const newState = storiesReducer(state, action);

    const expectedState = {
      data: [storyTwo],
      isLoading: false,
      isError: false
    };

    // toBe comparison will fail because it doesn't compare the content of the objects;
    // it is comparing the object reference. toStrictEqual will compare content instead
    expect(newState).toStrictEqual(expectedState);
  })
})

// ---------------------- UNIT TESTING: COMPONENTS ---------------------- //
describe('Item', () => {
  test('renders all properties', () => {
    render(<Item item={storyOne} />);

    // getByText grabs the element by the text and the "toBeInTheDoc.." and 
    // "toHaveAttribute" test if the characteristics are correct
    expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
    expect(screen.getByText('React')).toHaveAttribute('href', 'https://reactjs.org/')

    // Get in the habit of using React Testing Library's (RTL's) debug function
    // It gives a useful overview of what is rendered and informs best way to proceed
    // with testing
    screen.debug();
  });

  test('renders clickable "Dismiss" button', () => {
    render(<Item item={storyOne} />, () => {
      // getByRole retrieves by aria-label attributes
      expect(screen.getByRole('button')).toBeInTheDocument();
    });
  });

  /**
   * This tests that the button calls the callback handler.
   * It does NOT actually execute the logic of that handler
   */
  test('clicking the dismiss button calls the callback handler', () => {
    // This returns us a "mock" for the actual function and lets us capture when it's called
    const handleRemoveItem = jest.fn();

    render(<Item item={storyOne} onRemoveItem={handleRemoveItem} />);

    fireEvent.click(screen.getByRole('button'));

    expect(handleRemoveItem).toHaveBeenCalledTimes(1);
  })
});

describe('SearchForm', () => {
  const searchFormProps = {
    searchTerm: 'React',
    onSearchInput: jest.fn(),
    onSearchSubmit: jest.fn()
  };

  test('renders the input field with its value', () => {
    render(<SearchForm {...searchFormProps} />);

    expect(screen.getByDisplayValue('React')).toBeInTheDocument();

  });

  test('renders the correct label', () => {
    render(<SearchForm {...searchFormProps} />);

    // Using a basic regex to match 'Search', we can find the search text
    // without having to use the colon (Example: getByLabelText('Search:'))
    expect(screen.getByLabelText(/Search/)).toBeInTheDocument(); 
  });

  test('calls onSearchInput on input field change', () => {
    render(<SearchForm {...searchFormProps} />);

    fireEvent.change(screen.getByDisplayValue('React'),{
      target: {value: 'Redux'}
    });

    expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1);
  });

  test('calls searchSubmit on button submit click', () => {
    render(<SearchForm {...searchFormProps} />);

    fireEvent.submit(screen.getByRole('button'));
    
    expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
  })
})