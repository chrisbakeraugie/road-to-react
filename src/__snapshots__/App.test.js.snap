// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm renders snapshot 1`] = `
<form
  class="sc-gKAaRy jXtnzm"
>
  <div>
    <label
      for="search"
    >
      <strong>
        Search:
      </strong>
    </label>
    Â 
    <input
      id="search"
      type="text"
      value="React"
    />
  </div>
  <button
    class="sc-eCApnc sc-jSFjdj dysteu kzwYmg"
    type="submit"
  >
    Submit
  </button>
</form>
`;
